### PR TITLE
[FLINK-7923][Table API & SQL] Support field access of composite array element in SQL

### PR DESCRIPTION
**Pull request by suez1224**
Note: This is based on FLINK-7934, will rebase once FLINK-7934 is resolved.

## What is the purpose of the change

Support field access of composite array element in SQL. 

## Brief change log

  - add handling to calcite dot operator to support field access of composite array element in SQL
  - add unittests to verify that it works for tuple array, row array, pojo array and case class array

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unittests to verify the query plan*
  - *Added integration tests for batch/streaming for pojo/case class/tuple/row type*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by suez1224** [comment]
@twalthr rebased, could you please take another look? Thanks a lot.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/CalcTest.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+package org.apache.flink.table.api.stream.sql
+
+import org.apache.flink.api.scala._
+import org.apache.flink.table.api.scala._
+import org.apache.flink.table.utils.TableTestBase
+import org.apache.flink.table.utils.TableTestUtil._
+import org.junit.Test
+
+class CalcTest  extends TableTestBase {
```
Comment (hequn8128) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/CalcTest.scala]: Remove redundant blank between `CalcTest` and `extends`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 2]
```diff
@@ -984,6 +987,63 @@ object ScalarOperators {
     }
   }
 
+  def generateDot(codeGenerator: CodeGenerator,
```
Comment (hequn8128) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: Arguments of the method should be indented.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 3]
```diff
@@ -984,6 +987,63 @@ object ScalarOperators {
...
+      case crdt: CompositeRelDataType => {
+        val fieldName = dot.operands.get(1).asInstanceOf[RexLiteral]
+          .getValue.asInstanceOf[NlsString].getValue
+        if (crdt.compositeType.isInstanceOf[TupleTypeInfo[_]]) {
+           return GeneratedExpression(resultTerm, nullTerm,
+            s"""
+                   |${record.code}
+                   |${subField.code}
+                   |${resultTypeTerm} $resultTerm =
+                   |  (${resultTypeTerm}) ${record.resultTerm}.productElement(
```
Comment (hequn8128) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: NPE will be thrown if ${record.resultTerm} is null

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 4]
```diff
@@ -984,6 +987,63 @@ object ScalarOperators {
...
+                   |    ${fieldName.substring(1).toInt} - 1);
+                   |boolean $nullTerm =${resultTerm} == null;
+                   |""".stripMargin, resultType)
+        } else if (crdt.compositeType.isInstanceOf[CaseClassTypeInfo[_]]) {
+          return GeneratedExpression(resultTerm, nullTerm,
+            s"""
+               |${record.code}
+               |${resultTypeTerm} $resultTerm =
+               |  (${resultTypeTerm}) ${record.resultTerm}.${fieldName}();
```
Comment (hequn8128) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: NPE will be thrown

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 5]
```diff
@@ -984,6 +987,63 @@ object ScalarOperators {
...
+               |boolean $nullTerm =${resultTerm} == null;
+               |""".stripMargin, resultType)
+        } else if (crdt.compositeType.isInstanceOf[PojoTypeInfo[_]]) {
+          return GeneratedExpression(resultTerm, nullTerm,
+            s"""
+               |${record.code}
+               |${resultTypeTerm} $resultTerm =
+               |  (${resultTypeTerm}) ${record.resultTerm}.${fieldName};
```
Comment (hequn8128) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: NPE will be thrown

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 6]
```diff
@@ -984,6 +987,63 @@ object ScalarOperators {
...
+               |boolean $nullTerm =${resultTerm} == null;
+               |""".stripMargin, resultType)
+        } else if (crdt.compositeType.isInstanceOf[RowTypeInfo]) {
+          val fieldIndex = dot.operands.get(0).getType.asInstanceOf[CompositeRelDataType]
+            .compositeType.getFieldIndex(fieldName)
+          return GeneratedExpression(resultTerm, nullTerm,
+            s"""
+               |${record.code}
+               |${resultTypeTerm} $resultTerm =
+               |  (${resultTypeTerm}) ${record.resultTerm}.getField(${fieldIndex});
```
Comment (hequn8128) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: NPE will be thrown

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/SqlITCase.scala [hunk 7]
```diff
@@ -469,6 +473,148 @@ class SqlITCase extends StreamingWithStateTestBase {
...
 
+  @Test
+  def testArrayElementAtFromTableForTuple(): Unit = {
+
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
+    val tEnv = TableEnvironment.getTableEnvironment(env)
+    StreamITCase.clear
+
+    val data = List(
+      (1, Array((12, 45), (2, 5))),
```
Comment (hequn8128) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/SqlITCase.scala]: 1. Add null tuple input test
2. Add nested tuple input test

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/CalcTest.scala [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  def testArrayElement(): Unit = {
+    val util = streamTestUtil()
+    util.addTable[(Long, Array[(String, Int)])]("MyTable", 'a, 'b)
+
+    val expected = unaryNode(
+      "DataStreamCalc",
+      streamTableNode(0),
+      term("select",
+        "a",
+        "DOT(ITEM(b, 1), '_1') AS b11"
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/CalcTest.scala]: Actually we don't need a full table test for this. As it only tests a scalar operation. Can you move the test to `CompositeAccessTest` and test it for all APIs.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala (hunk 2 above)
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: I think this method is not really necessary. We already have logic for accessing composite types (see `visitFieldAccess()`). Maybe we just have to make the methods there a bit more generic.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/SqlITCase.scala (hunk 7 above)
Comment (suez1224) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/SqlITCase.scala]: Added null check.
for nested tuple input, it wont work for now due to https://issues.apache.org/jira/browse/CALCITE-2162. I've submitted a fix to it, should be available in Calcite 1.16.

**Comment by suez1224** [comment]
@hequn8128 , thanks for the review. Where do you think we can add it in sql.md?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala (hunk 2 above)
Comment (suez1224) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: In Calcite, once it sees array element access, the subsequent field access is translated into DOT RexCall, not RexFieldAccess. Therefore, we need to a custom handling for the DOT RexCall.

Comment (suez1224) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: But I agree I should refactor the code in visitFieldAccess() to reuse it as much as possible.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/CalcTest.scala (hunk 8 above)
Comment (suez1224) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/sql/CalcTest.scala]: Will fix it.

**Comment by twalthr** [comment]
Thanks for the update @suez1224. The changes look good. I will merge this...

**Comment by twalthr** [comment]
Btw I will remove the null tests because fields of tuples and case classes are not allowed to be null. The serializers would throw an exception.

**Pull request closed without merging**